src/main/kotlin/org/rust/ide/annotator/RustLiteralAnnotator.kt

```diff
@@ -0,0 +1,80 @@
+package org.rust.ide.annotator
+
+import com.intellij.lang.annotation.Annotation
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.Annotator
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import org.rust.lang.core.psi.RustLiteral
+import org.rust.lang.core.psi.RustTokenElementTypes.*
+import org.rust.lang.core.psi.visitors.RustVisitorEx
+
+class RustLiteralAnnotator : Annotator {
+    override fun annotate(element: PsiElement, holder: AnnotationHolder) = element.accept(object : RustVisitorEx() {
+        override fun visitLiteral(literal: RustLiteral) {
+            // Check suffix
+            val suffix = literal.suffix
+            val possibleSuffixes = literal.possibleSuffixes
+            if (!suffix.isNullOrEmpty() && suffix !in possibleSuffixes) {
+                holder.literalError(literal, if (possibleSuffixes.isNotEmpty()) {
+                    val possibleSuffixesStr = possibleSuffixes.map { "'$it'" }.joinToString()
+                    "invalid suffix '$suffix' for ${literal.displayName}; " +
+                        "the suffix must be one of: $possibleSuffixesStr"
+                } else {
+                    "${literal.displayName} with a suffix is invalid"
+                })
+            }
+        }
+
+        override fun visitNumericLiteral(literal: RustLiteral.Number) {
+            // TODO: Check numeric literals boundaries
+            super.visitNumericLiteral(literal)
+        }
+
+        override fun visitTextLiteral(literal: RustLiteral.Text) {
+            // Check char literal length
+            when (literal.tokenType) {
+                BYTE_LITERAL, CHAR_LITERAL -> {
+                    val value = literal.value
+                    val message = when {
+                        value == null || value.length == 0 -> "empty ${literal.displayName}"
+                        value.length > 1                   -> "too many characters in ${literal.displayName}"
+                        else                               -> null
+                    }
+                    if (message != null) holder.literalError(literal, message)
+                }
+            }
+
+            // Check delimiters
+            if (literal.hasUnpairedQuotes) {
+                holder.literalError(literal, "unclosed ${literal.displayName}")
+            }
+
+            super.visitTextLiteral(literal)
+        }
+    })
```

Comment:
I have an alternative design for implementing this. This whole visitor could be replaced with a simple invocation of `RustLiteral.validate(): Collection<String>` method (or even `RustLiteral.validate(): Collection<RustLiteral.Error>`) where the result is a collection of error messages (errors) (ofc the annotator will iterate over them and yield error annotations).

Such design will enable easy implementation of `isValid` property, but it will mix literal validation code with it's PSI impl.